[Quant] Add fused LinearTanh module for onednn backend

Xia-Weiwen opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89188
* __->__ #88923
* #88879
* #88668


**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Xia-Weiwen committed:
[Quant] Add fused LinearTanh module for onednn backend

[ghstack-poisoned]

jerryzh168 reviewed (commented):
looks good overall, but probably need some update after the previous PRs are updated I think

jgong5 reviewed (approved):
Not sure if we need to rename the files to something like linear_activation from linear_relu since we added linear+tanh into it.

Xia-Weiwen commented:
> Not sure if we need to rename the files to something like linear_activation from linear_relu since we added linear+tanh into it.

Good idea. Thanks. I will do that after clarifications from Jerry about PT 2.0.

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"

**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule




[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168. Is it ok to land this? Thanks

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
Hi @jerryzh168 Do you have more comments on this PR? Thanks!

jerryzh168 commented on torch/nn/intrinsic/__init__.py:
```diff
@@ -12,6 +12,7 @@
 from torch.ao.nn.intrinsic import BNReLU3d
 from torch.ao.nn.intrinsic import LinearBn1d
 from torch.ao.nn.intrinsic import LinearLeakyReLU
+from torch.ao.nn.intrinsic import LinearTanh
```
please revert these, we don't want to add new things to torch/nn/intrinsic folder, this is a folder that we plan to deprecate in the future

Xia-Weiwen replied:
Ok they are reverted.

jerryzh168 reviewed (changes requested):
please revert changes to torch/nn/intrinsic folder

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
> please revert changes to torch/nn/intrinsic folder

Ok they are reverted.

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jerryzh168 commented on test/allowlist_for_publicAPI.json:
```diff
@@ -12,7 +12,7 @@
     "torch.nn.intrinsic.quantized.modules.bn_relu": "torch.ao.nn.intrinsic.quantized.modules.bn_relu",
     "torch.nn.intrinsic.quantized.modules.conv_relu": "torch.ao.nn.intrinsic.quantized.modules.conv_relu",
-    "torch.nn.intrinsic.quantized.modules.linear_relu": "torch.ao.nn.intrinsic.quantized.modules.linear_relu",
+    "torch.nn.intrinsic.quantized.modules.linear_activation": "torch.ao.nn.intrinsic.quantized.modules.linear_activation",
```
is this change still valid? we don't have `torch.nn.intrinsic.quantized.modules.linear_activation` now

Xia-Weiwen replied:
It's still needed whereas the file name id changed:
```
"torch.nn.intrinsic.quantized.modules.linear_relu": "torch.ao.nn.intrinsic.quantized.modules.linear_activation",
```

Xia-Weiwen replied:
I reverted the file name change due to CI checks failure.

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen commented:
> > > Not sure if we need to rename the files to something like linear_activation from linear_relu since we added linear+tanh into it.
> > 
> > 
> > Good idea. Thanks. I will do that after clarifications from Jerry about PT 2.0.
> 
> I have renamed the file to `linear_activation.py`

I reverted the change due to CI check failure.

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Xia-Weiwen committed:
Update on "[Quant] Add fused LinearTanh module for onednn backend"



**Summary**
This PR adds fused `QLinearTanh` module for onednn backend, which will be used for int8 inference with onednn backend. Cannot call this module with other quantization backends otherwise an error is thrown.

**Test plan**
python test_quantization.py TestStaticQuantizedModule

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]
